walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,117 @@
+    Event lastEvent = eventList.get(eventList.size() - 1);
+
+    int eventsSkipped = 0;
+
+    if(eventList.size() == 1) {
+      if(!requestHasAttendees(request, firstEvent)) {
+        eventsSkipped++;
+      }
+    }
+
+    // Add the first time slot, from the start of the day to the start of the first event,
+    // only if there's no event that starts the day.
+    if(dayStartsWithEvent(eventList) && requestHasAttendees(request, firstEvent)) {
+      availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+        eventList.get(0).getWhen().start(), false));
+    }
+
+    for (int i = 0; i < eventList.size() - 1; i++) {
+      TimeRange currEventTimeRange = eventList.get(i).getWhen();
+      TimeRange nextEventTimeRange = eventList.get(i + 1).getWhen();
+
+      if(!requestHasAttendees(request, currEvent)) {
+        eventsSkipped ++;
+        continue;
+      }
+
+      if(currEventTimeRange.contains(nextEventTimeRange)) {
+        // If there is a 3rd event, connect events 1 and 3. Otherwise, connect with the end of day.
+        if(i + 2 < eventList.size() - 1) {
+          TimeRange nextnextEventTimeRange = eventList.get(i + 2).getWhen();
+          availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+            nextnextEventTimeRange.start(), true));
+          i++;
+        } else {
+          availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+            TimeRange.END_OF_DAY, true));
+        }
+      } else {
+          if(currEventTimeRange.overlaps(nextEventTimeRange)) {
+            continue;
+          }
+
+          if(nextEventTimeRange.start() - currEventTimeRange.end() >= request.getDuration()) {
+            availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+              nextEventTimeRange.start(), false));  
+          } 
+        }
+    }
+
+    if(requestHasAttendees(request, lastEvent)) {
+      if(availableTimes.size() != 0){
+        if(nothingEndsTheDay(availableTimes, eventList)) {
+          availableTimes.add(TimeRange.fromStartEnd(eventList.get(eventList.size() - 1).getWhen().end(),
+            TimeRange.END_OF_DAY, true));
+        } 
+      } else {
+          if(fitsOnlyAtTheEnd(request, availableTimes, eventList)) {
+              availableTimes.add(TimeRange.fromStartEnd(eventList.get(eventList.size() - 1).getWhen().end(),
+                TimeRange.END_OF_DAY, true));
+            }
+        } 
+    } 
+
+    if(eventsSkipped == eventList.size()) {
+      availableTimes.add(TimeRange.WHOLE_DAY);
+    }
+
+      return availableTimes;
+  }
+
+  private boolean requestHasAttendees(MeetingRequest request, Event currEvent) {
+    List<String> requestAttendees = new ArrayList<>(request.getAttendees());
+    List<String> eventAttendees = new ArrayList<>(currEvent.getAttendees());
+
+    List<String> commonAttendees = requestAttendees.stream()
+      .filter(eventAttendees::contains)
+      .collect(Collectors.toList());
```

Comment:
You don't need to create the full list of commonAttendees, as you only want to know if there is at least one attendee in common. You can for example use anyMatch instead of filter, which will be more efficient as it will stop once it found the first match